kubeflow/examples/prototypes/tf-batch-predict.jsonnet

```diff
@@ -0,0 +1,193 @@
+// @apiVersion 0.1
+// @name io.ksonnet.pkg.tf-batch-predict
+// @description TensorFlow batch-predict
+// @shortDescription A TensorFlow batch-predict job
+// @param name string Name to give to each of the components
+// @optionalParam numGpus number 0 number of GPUs to use
+// @param modelPath string 0 Path to the model directory
+// @param inputFilePatterns string Input file patterns
+// @param outputResultPrefix string Output result file prefix
+// @param outputErrorPrefix string Output error file prefix
+// @optionalParam batchSize number 8 Batch size
+// @optionalParam gcpCredentialSecretName string Secret name if used in GCP
+
+local k = import "k.libsonnet";
+
+local tfBatchPredictBase = {
+  // Parameters are intended to be late bound.
+  local base = self,
+  params:: {
+    name: null,
+    labels: {
+      app: base.params.name,
+    },
+    modelName: self.name,
+    modelPath: null,
+    inputFilePatterns: null,
+    inputFileFormat: null,
+    outputResultPrefix: null,
+    outputErrorPrefix: null,
+    batchSize: 8,
+    numGpus: 0,
+
+    version: "v1",
+
+    // If users want to override the image then can override defaultCpuImage and/or defaultGpuImage
+    // in which case the image used will still depend on whether GPUs are used or not.
+    // Users can also override the predictImage in which case the user supplied value will always be used
+    // regardless of numGpus.
+    defaultCpuImage: "gcr.io/kubeflow-examples/batch-predict:tf18",
+    defaultGpuImage: "gcr.io/kubeflow-examples/batch-predict:tf18-gpu",
+
+    predictImage: if self.numGpus == 0 then
+      self.defaultCpuImage
+    else
+      self.defaultGpuImage,
+
+    // Which cloud to use
+    cloud:: "gcp",
+  },
+
+  // Parametes specific to GCP.
+  gcpParams:: {
+    gcpCredentialSecretName: "",
+  } + self.params,
+
+
+  parts:: {
+    // We define the containers one level beneath parts because combined with jsonnet late binding
+    // this makes it easy for users to override specific bits of the container.
+    tfBatchPredictContainerBase:: {
+      name: base.params.name,
+      image: base.params.predictImage,
+      imagePullPolicy: "IfNotPresent",
+      args: [
+        "--model_dir=" + base.params.modelPath,
+        "--input_file_patterns=" + base.params.inputFilePatterns,
+        "--input_file_format=" + base.params.inputFileFormat,
+        "--output_result_prefix=" + base.params.outputResultPrefix,
+        "--output_error_prefix=" + base.params.outputErrorPrefix,
+        "--batch_size=" + base.params.batchSize,
+      ],
+      resources: {
+        requests: {
+          memory: "4Gi",
+          cpu: "1",
+        },
+        limits: {
+          memory: "4Gi",
+          cpu: "4",
+        },
+      },
+    },  // tfBatchPredictContainerBase
+
+    tfBatchPredictContainer+: base.parts.tfBatchPredictContainerBase +
```

Comment:
Lets simplify the prototype further. There's not much reason to pull out tfBatchPredictContainer and then override by merging in a dictionary; just paste the relevant dictionary describing the container into the container arrays below line 111.

We want the examples to be optimized for readability. The idea is that if people want to modify them (e.g. they want to set appropriate keys for AWS they would delete and replace the appropriate lines.

In other words, think of this as an "example" that people will copy and modify.

Ideally the example should run under some suitable set of conditions; e.g. assuming there is a valid GCP credential either as a VM service account or as a secret key if secretName is set.

I think we should keep conditionals for setting the following based on parameters

1. num GPUs
2. GCP secret.